### PR TITLE
ci: refactor single-gpu e2e_tests

### PR DESCRIPTION
**Pull request by MikhailKardash**
## Description

This PR is part of a refactoring of ML Systems related e2e_tests. 
 - Refactor tests that do not require multiple GPU runners.
 - Set up GPU runners in CircleCI and use docker to perform runs

## Test Plan

Run CI unit tests:
- test_unit/test-unit-harness-gpu
- test_unit/test-unit-harness-gpu-lightning

## Commentary (optional)

CircleCI GPU runners are currently using docker directly, but should be containerized in the future. 
Multi-GPU runners are still a mystery. 
There are some GPU unit tests that are currently broken as they were never maintained--updating those is currently a separate ticket as they are currently disabled anyway.
Pytest tagging and test directory structure may need to change going forward. "not gpu" and "gpu" tagging is not very robust.

## Checklist

- [X] Changes have been manually QA'd

## Ticket

MLG-488
MLG-479

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -410,6 +410,29 @@ commands:
...
+      extras-requires:
+        type: string
+        default: ""
+    steps:
+      - run: docker run -v <<parameters.bind-mount>>:/usr/src/project --gpus=all -it -d --name <<parameters.image-name>> --workdir /usr/src/project <<parameters.image>>
+      - run:
+          name: Install <<parameters.extras-requires>>
+          command: |
+            if [ -n "<<parameters.extras-requires>>" ]; then
+              docker exec <<parameters.image-name>> tools/scripts/retry.sh pip install --progress-bar off <<parameters.extras-requires>>
```
Comment (rb-determined-ai) [inline .circleci/config.yml:433]: I would _love_ to see us get rid of all test-time pip installs...  But I do think it is wise to do that in a separate PR.

File: .circleci/config.yml [hunk 2]
```diff
@@ -410,6 +410,29 @@ commands:
...
+        type: string
+        default: "test_container"
+      bind-mount:
+        type: string
+        default: "$PWD"
+      extras-requires:
+        type: string
+        default: ""
+    steps:
+      - run: docker run -v <<parameters.bind-mount>>:/usr/src/project --gpus=all -it -d --name <<parameters.image-name>> --workdir /usr/src/project <<parameters.image>>
```
Comment (rb-determined-ai) [inline .circleci/config.yml:428]: _This is kind of a nit, but it'll get weird if we choose this pattern now and stick to it everywhere, so..._

Let's stick to the [Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs.html) when picking file locations.  The `/usr` directory is meant to be [read-only](ttps://refspecs.linuxfoundation.org/FHS_3.0/fhs.html#purpose18) in most cases, so it's a weird place ot mount this stuff.

Instead, let's put everything in `/mnt`, which is meant for [temporarily mounted files](https://refspecs.linuxfoundation.org/FHS_3.0/fhs.html#mntMountPointForATemporarilyMount), which is closer to what we're doing here.

File: harness/Makefile [hunk 3]
```diff
@@ -28,7 +28,13 @@ test-cpu:
 .PHONY: test-gpu
 test-gpu:
 	# We will use this target in circleci once we have access to gpu instances.
```
Comment (rb-determined-ai) [inline harness/Makefile:30]: update comment?

File: harness/tests/experiment/integrations/test_pytorch_lightning.py [hunk 4]
```diff
@@ -14,6 +15,8 @@
 from tests.experiment.fixtures import lightning_adapter_onevar_model as la_model
 
 
+@pytest.mark.PyTorchLightning
```
Comment (rb-determined-ai) [inline harness/tests/experiment/integrations/test_pytorch_lightning.py:18]: can we stick to `snake_case` for our tag names?

File: harness/tests/experiment/integrations/test_pytorch_lightning.py [hunk 5]
```diff
@@ -14,6 +15,8 @@
...
+@pytest.mark.gpu
```
Comment (rb-determined-ai) [inline harness/tests/experiment/integrations/test_pytorch_lightning.py:19]: why is the whole class GPU-only?

File: harness/tests/experiment/keras/test_tf_keras_trial.py [hunk 6]
```diff
@@ -365,6 +365,133 @@ def make_workloads() -> workload.Stream:
         controller.run()
 
 
+# TODO: maybe add TF2 and TF1 images with pytest.mark (We're dropping TF1 images anyway)
+@pytest.mark.TensorFlow
```
Comment (rb-determined-ai) [inline harness/tests/experiment/keras/test_tf_keras_trial.py:369]: `snake_case` tags again

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 7]
```diff
@@ -136,7 +136,7 @@ def test_checkpointing_and_restoring(self, tmp_path: pathlib.Path) -> None:
         self.checkpoint_and_restore(updated_hparams, tmp_path, (100, 100))
 
     @pytest.mark.skipif(not torch.cuda.is_available(), reason="no gpu available")
-    @pytest.mark.gpu
+    # @pytest.mark.gpu
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:139]: what's going on here?

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 8]
```diff
@@ -919,6 +919,51 @@ def checkpoint_and_restore(
 
         return (training_metrics["A"], training_metrics["B"])
 
+    def checkpoint_and_restore_no_callbacks(
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:922]: I don't think I understand what "_no_callbacks" means in the name here?

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 9]
```diff
@@ -919,6 +919,51 @@ def checkpoint_and_restore(
...
+            min_checkpoint_batches=steps[0],
+            checkpoint_dir=checkpoint_dir,
+            expose_gpus=True,
+        )
+
+        trial_controller_A.run()
+
+        assert len(os.listdir(checkpoint_dir)) == 1, "trial did not create a checkpoint"
+
+        # Trial A: restore from checkpoint and train for 100 more batches
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:949]: rename to trial_B?

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 10]
```diff
@@ -957,6 +1002,73 @@ def test_legacy_checkpoint_loading(self, ckpt: str):
         assert state.batches_trained == 1, "batches_trained does not match"
         assert state.epochs_trained == 0, "epochs_trained does not match"
 
+    @pytest.mark.gpu
+    @pytest.mark.parametrize("aggregation_frequency", [1, 4])
+    def test_pytorch_11_const(self, aggregation_frequency: int, tmp_path: pathlib.Path):
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:1007]: What is the purpose of this test?

The old e2e test looked like it was meant to test multiple images, but if we're doing unit tests against multiple images, what coverage does this particular test offer that we wouldn't get by running other tests in multiple images?

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 11]
```diff
@@ -957,6 +1002,73 @@ def test_legacy_checkpoint_loading(self, ckpt: str):
...
+            hparams=hparams,
+            tmp_path=tmp_path,
+            exp_config=exp_config,
+            steps=(1, 1),
+        )
+
+    @pytest.mark.gpu
+    @pytest.mark.PyTorch
+    @pytest.mark.parametrize("api_style", ["apex", "auto", "manual"])  # TODO: test apex
+    def test_pytorch_const_with_amp(self, api_style: str, tmp_path: pathlib.Path):
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:1038]: what does this test offer that our other gpu amp unit tests don't?

I suspect, but don't recall exactly, that we wrote the old unit tests with the intention of turning them on in the future, and the e2e tests were just to have some coverage of the feature in the mean time.

File: harness/tests/experiment/utils.py [hunk 12]
```diff
@@ -174,6 +177,61 @@ def repo_path(path: str) -> str:
     return os.path.join(os.path.dirname(__file__), "../../../", path)
 
 
+def e2e_fixtures_path(path: str) -> str:
+    # This function is a helper for testing, all necessary fixtures should be refactored instead.
+    # TODO: refactor fixtures.
```
Comment (rb-determined-ai) [inline harness/tests/experiment/utils.py:182]: I'd recommend copying the fixtures you need to here, and letting the fixtures in e2e get deleted over time, rather than introduce a sneaky cross-repo dependency.

But only if we're pretty sure the fixtures in question _will_ get deleted from e2e in a reasonably short timeframe, I guess.  If they're gonna be kept for a while, probably that's bad advice...

File: harness/tests/experiment/utils.py [hunk 13]
```diff
@@ -174,6 +177,61 @@ def repo_path(path: str) -> str:
...
+
+def gan_examples_path(path: str) -> str:
+    return os.path.join(os.path.dirname(__file__), "../../../examples/gan", path)
+
+
+def tutorials_path(path: str) -> str:
+    return os.path.join(os.path.dirname(__file__), "../../../examples/tutorials", path)
+
+
+def import_module(module_name: str, module_path: str, model_context: Optional[str] = None) -> Any:
```
Comment (rb-determined-ai) [inline harness/tests/experiment/utils.py:198]: why not use the shiny new `det.import_from_path` that I wrote?

This is basically what it's designed for.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 7 above)
Comment (MikhailKardash) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:139]: We have some GPU tests that were not maintained after changes to our trial context/controller objects. I created a ticket for these, MLG-479. I figure enabling these is out of the scope of this PR, so I kept them disabled. If we decide that these can replace some e2e_tests, then it may be worth fixing.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 8 above)
Comment (MikhailKardash) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:922]: there is a checkpoint_and_restore function that uses callback methods that are tied to pytorch_onevar_model. Wanted to make it clear in the code why there needs to be a second function, but I should probably add that as a dev note instead.

File: harness/tests/experiment/utils.py (hunk 12 above)
Comment (MikhailKardash) [reply harness/tests/experiment/utils.py:182]: This function is here just for prototyping new unit tests before I refactor any fixtures. It's not actually used in the final code, and will be removed after this refactor is complete.

Comment (MikhailKardash) [reply harness/tests/experiment/utils.py:182]: Fixtures are being moved here along with their tests. This function is here to test code before I do any refactoring as a sanity check. I plan to delete it in the future along with the last refactor-related PR.

**Review by loksonarius — APPROVED** [review]
Approving circle config changes

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 10 above)
Comment (MikhailKardash) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:1008]: At the very least, it tests the MNIST Pytorch example code. There's also a config parameter that is changed.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 11 above)
Comment (MikhailKardash) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:1038]: We can remove this if the e2e_tests were only a temporary solution, since `test_scaler_checkpointing_and_restoring` does the same thing but with extra metrics checking. Just need to fix the tests.

File: harness/tests/experiment/fixtures/pytorch_onevar_model.py [hunk 14]
```diff
@@ -336,15 +336,26 @@ class OneVarAMPBaseTrial(OneVarTrial):
...
+        skip_batches = self.context.get_initial_batch()
+
+        sampler = torch.utils.data.SequentialSampler(dataset)
+        sampler = samplers.ReproducibleShuffleSampler(sampler, seed)
+        sampler = samplers.RepeatSampler(sampler)
+        sampler = samplers.DistributedSampler(sampler, num_workers=num_workers, rank=rank)
+        batch_sampler = torch.utils.data.BatchSampler(sampler, batch_size, drop_last=False)
+        batch_sampler = samplers.SkipBatchSampler(batch_sampler, skip_batches)
+
+        return pytorch.DataLoader(dataset, batch_sampler=batch_sampler)
```
Comment (rb-determined-ai) [inline harness/tests/experiment/fixtures/pytorch_onevar_model.py:358]: what's the goal of using a custom sampler pipeline, but then feeding the resulting sampler into our pytorch.DataLoader (which will add another skip part to the pipeline)?

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 15]
```diff
@@ -1104,7 +1229,7 @@ def create_trial_and_trial_controller(
             exp_conf=exp_config,
             aggregation_frequency=1,
             steps_completed=steps_completed,
-            managed_training=False,
+            managed_training=True,  # dev note: this must be True to put model on GPU
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:1232]: all comments are dev notes, you don't need to say `dev note:`

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 10 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:1020]: OK, I'm down with testing the example.

But what is the value of testing aggregation frequency with the example?

I like unit tests to be separated into testing specific things.  We want test coverage of features:
- aggregation frequency
- multiple versions of torch
- example code

I don't like mixing those very much.  Testing example code should be focused on making sure examples work, and nothing else.  Some day, if we delete the example, we should delete the example tests, and it shouldn't affect the coverage of our core features.

So it seems like we already test aggregation frequency elsewhere, right?  So are we safe to remove aggregation frequency from our example test?

Also, why do we need to test our mnist example with an actual GPU?  Isn't the cpu test coverage good enough for the mnist model?

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 16]
```diff
@@ -911,6 +921,9 @@ def checkpoint_and_restore(
         training_metrics["B"] = metrics_callback.training_metrics
         validation_metrics["B"] = metrics_callback.validation_metrics
 
+        print(training_metrics)
+        print(validation_metrics)
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:925]: don't forget to remove stray prints before landing.

File: harness/tests/experiment/keras/test_tf_keras_trial.py [hunk 17]
```diff
@@ -365,6 +365,131 @@ def make_workloads() -> workload.Stream:
         controller.run()
 
 
+@pytest.mark.tensorflow
+@pytest.mark.gpu
```
Comment (rb-determined-ai) [inline harness/tests/experiment/keras/test_tf_keras_trial.py:369]: does cifar10 need gpu testing?  Why can't we run the cifar model against just CPUs and call it a day?

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 10 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:1020]: side note, we actually have dedicated example testing in `examples/tests` I think

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 11 above)
Comment (MikhailKardash) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:1049]: removed

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 10 above)
Comment (MikhailKardash) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py]: Removed pytorch mnist and tensorflow cifar10 examples from examples/tests since the new tests are better.

File: harness/tests/experiment/fixtures/pytorch_onevar_model.py [hunk 18]
```diff
@@ -417,6 +415,7 @@ def train_batch(
         # Measure the weight right now.
         w_before = self.model.weight.data.item()
 
+        # XXXXXX LOOK AT THIS AS WELL XXXXXXXXX
```
Comment (rb-determined-ai) [inline harness/tests/experiment/fixtures/pytorch_onevar_model.py:418]: ...

File: harness/tests/test_imports.py [hunk 19]
```diff
@@ -129,3 +129,6 @@ def chdir(path: pathlib.Path) -> Iterator:
     # We will have created some __pycache__ dirs but we don't want to pollute.
     shutil.rmtree(fixture / "a" / "__pycache__")
     shutil.rmtree(fixture / "libraries" / "__pycache__")
+
+    for mod in ["lib1", "lib2", "data"]:
```
Comment (rb-determined-ai) [inline harness/tests/test_imports.py:133]: Worth a comment explaining why.

File: harness/tests/experiment/utils.py [hunk 20]
```diff
@@ -174,6 +176,41 @@ def repo_path(path: str) -> str:
     return os.path.join(os.path.dirname(__file__), "../../../", path)
 
 
+def e2e_fixtures_path(path: str) -> str:
+    # This function is a helper for testing, all necessary fixtures should be refactored instead.
+    return os.path.join(os.path.dirname(__file__), "../../../e2e_tests/tests/fixtures", path)
```
Comment (rb-determined-ai) [inline harness/tests/experiment/utils.py:181]: I thought you were going to address this hack before landing?

File: e2e_tests/tests/experiment/test_pytorch.py [hunk 21]
```diff
@@ -7,116 +7,6 @@
 from tests import experiment as exp
 
 
-@pytest.mark.e2e_gpu
-@pytest.mark.parametrize("aggregation_frequency", [1, 4])
-@pytest.mark.parametrize("image_type", ["PT", "TF2"])
```
Comment (rb-determined-ai) [inline e2e_tests/tests/experiment/test_pytorch.py:12]: last comment: are we replacing this two-image test functionality somewhere in the new tests?

Comment (MikhailKardash) [reply e2e_tests/tests/experiment/test_pytorch.py:12]: No, but we still have coverage of this in other e2e_tests (see `test_pytorch_cifar10_parallel` ). If we want to test the image, I would make a pytorch-image runner with a selected suite of tests.

Comment (rb-determined-ai) [reply e2e_tests/tests/experiment/test_pytorch.py:12]: if we have the coverage in e2e under another test, I'm ok with that, and we can land this as-is.

Before we delete that _that_ test though, we need to bring that coverage into unit tests as well.

**Review by rb-determined-ai — APPROVED** [review]
awesome, dude
